sys/include/usb/usbus.h

```diff
@@ -0,0 +1,533 @@
+     * event handler
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     */
+    void (*init)(usbus_t * usbus, struct usbus_handler *handler);
+
+    /**
+     * @brief event handler function
+     *
+     * This function is passed USBUS events
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     * @param event     @ref USBUS_MSG_TYPES "event" to handle
+     */
+    void (*event_handler)(usbus_t * usbus, struct usbus_handler *handler,
+                          usbus_event_usb_t event);
+
+    /**
+     * @brief transfer handler function
+     *
+     * This function receives transfer based events
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     * @param ep        usbdev endpoint that triggered the event
+     * @param event     @ref usbus_event_transfer_t event
+     */
+    void (*transfer_handler)(usbus_t * usbus, struct usbus_handler *handler,
+                             usbdev_ep_t *ep, usbus_event_transfer_t event);
+
+    /**
+     * @brief setup request handler function
+     *
+     * This function receives USB setup requests from the USBUS stack.
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     * @param state     setup request state
+     * @param setup     setup packet
+     *
+     * @returns         Size of the returned data when the request is handled
+     * @returns         negative to have the stack return an USB stall to the
+     *                  host
+     * @returns         zero when the request is not handled by this handler
+     */
+    int (*setup_handler)(usbus_t * usbus, struct usbus_handler *handler,
+                         usbus_setuprq_state_t state, usb_setup_t *request);
+} usbus_handler_driver_t;
+
+/**
+ * @ USBUS handler struct
+ *
+ * Inherit from this struct for custom USB functionality
+ */
+struct usbus_handler {
+    struct usbus_handler *next;             /**< List of handlers (to be used by
+                                                 @ref usbus_t) */
+    const usbus_handler_driver_t *driver;   /**< driver for this handler */
+    usbus_interface_t *iface;               /**< Interface this handler belongs
+                                                 to */
+    uint32_t flags;                         /**< Report flags */
+};
+
+/**
+ * @brief USBUS context struct
+ */
+struct usbus {
+    usbus_string_t manuf;                           /**< Manufacturer string                   */
+    usbus_string_t product;                         /**< Product string                        */
+    usbus_string_t config;                          /**< Configuration string                  */
+    usbus_endpoint_t ep_out[USBDEV_NUM_ENDPOINTS];  /**< USBUS OUT endpoints                   */
+    usbus_endpoint_t ep_in[USBDEV_NUM_ENDPOINTS];   /**< USBUS IN endpoints                    */
```

Comment:
One "big" change with the latest fixup is that the `usbus_endpoint_t`'s are now statically allocated within the usbus context struct. This effectively moves them from each handler to the central context. A number equal to the usbdev endpoints is allocated, effectively matching them.

Reply:
The motivation for this change lies within the change to thread flags. As a thread flag doesn't carry any content, a different way to grab the endpoint that requires servicing was necessary. The solution is a `uint32_t` acting as a holder for bitflags, one for each endpoint (USB has max 16 IN and 16 OUT endpoints as protocol limitation). From this bitmap, the endpoint direction and number that requires servicing is retrieved. The usbus endpoint context and from there the usbdev endpoint context is then retrieved from this array.

A different solution more in line with the previous decentral endpoint context storage would be to iterate all interfaces and for each interface check if it contains the endpoint that we're looking for. As this requires a lot of iterating, I decided that centralizing this storage and potentially allocated too many endpoint structs is the more favorable solution here.